Sources/SwiftFormatCore/RuleMask.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+/// This class takes the raw source text and scans through it searching for comment pairs of the
+/// form:
+///
+///   3. |  // swift-format-disable: RuleName
+///   4. |  let a = 123
+///   5. |  // swift-format-enable: RuleName
+///
+/// This class records that `RuleName` is disabled for line 4. The rules themselves reference
+/// RuleMask to see if it is disabled for the line it is currently examining.
+public class RuleMask {
+
+  /// Each rule has a list of ranges for which it is disabled.
+  private var ruleMap: [String: [Range<Int>]] = [:]
+
+  /// Regex to match the enable comments; rule name is in the first capture group.
+  private let enablePattern = #"\s*//\s*swift-format-enable:\s+(\S+)"#
```

Comment:
I totally forgot that we've had raw strings since Swift 5 😄

Reply:
This has the potential to catch things we don't want, like:

```swift
let howToDisable = "Add // swift-format-enable: foo to your source file"
```

Even if we required things like putting the comment on its own line, multi-line strings wouldn't be safe. So I'm starting to wonder if the token-based approach would be better...

I think it's still worth checking this in, since the overall infrastructure would be the same, but after that I think we should look at an approach that walks the tokens of the file and looks for the comments in the trivia rather than just doing line-based comparisons.

Reply:
I updated the regex to defend against the case you pointed out. However, it will still be vulnerable to multiline string literals.